crates/bevy_ecs/src/query/fetch.rs

```diff
@@ -0,0 +1,818 @@
+/// SAFETY:
+/// Implementor must ensure that [FetchState::update_component_access] and [FetchState::update_archetype_component_access] exactly
+/// reflects the results of [FetchState::matches_archetype], [FetchState::matches_table], [Fetch::archetype_fetch], and [Fetch::table_fetch]
+pub unsafe trait FetchState: Send + Sync + Sized {
+    fn init(world: &mut World) -> Self;
+    fn update_component_access(&self, access: &mut FilteredAccess<ComponentId>);
+    fn update_archetype_component_access(
+        &self,
+        archetype: &Archetype,
+        access: &mut Access<ArchetypeComponentId>,
+    );
+    fn matches_archetype(&self, archetype: &Archetype) -> bool;
+    fn matches_table(&self, table: &Table) -> bool;
+}
+
+/// A fetch that is read only. This should only be implemented for read-only fetches.
+pub unsafe trait ReadOnlyFetch {}
+
+impl WorldQuery for Entity {
+    type Fetch = EntityFetch;
+    type State = EntityState;
+}
+
+pub struct EntityFetch {
+    entities: *const Entity,
+}
+
+/// SAFE: access is read only
+unsafe impl ReadOnlyFetch for EntityFetch {}
+
+pub struct EntityState;
+
+// SAFE: no component or archetype access
+unsafe impl FetchState for EntityState {
+    fn init(_world: &mut World) -> Self {
+        Self
+    }
+
+    fn update_component_access(&self, _access: &mut FilteredAccess<ComponentId>) {}
+
+    fn update_archetype_component_access(
+        &self,
+        _archetype: &Archetype,
+        _access: &mut Access<ArchetypeComponentId>,
+    ) {
+    }
+
+    #[inline]
+    fn matches_archetype(&self, _archetype: &Archetype) -> bool {
+        true
+    }
+
+    #[inline]
+    fn matches_table(&self, _table: &Table) -> bool {
+        true
+    }
+}
+
+impl<'w> Fetch<'w> for EntityFetch {
+    type Item = Entity;
+    type State = EntityState;
+
+    #[inline]
+    fn is_dense(&self) -> bool {
+        true
+    }
+
+    unsafe fn init(_world: &World, _state: &Self::State) -> Self {
+        Self {
+            entities: std::ptr::null::<Entity>(),
+        }
+    }
+
+    #[inline]
+    unsafe fn set_archetype(
+        &mut self,
+        _state: &Self::State,
+        archetype: &Archetype,
+        _tables: &Tables,
+    ) {
+        self.entities = archetype.entities().as_ptr();
+    }
+
+    #[inline]
+    unsafe fn set_table(&mut self, _state: &Self::State, table: &Table) {
+        self.entities = table.entities().as_ptr();
+    }
+
+    #[inline]
+    unsafe fn table_fetch(&mut self, table_row: usize) -> Self::Item {
+        *self.entities.add(table_row)
+    }
+
+    #[inline]
+    unsafe fn archetype_fetch(&mut self, archetype_index: usize) -> Self::Item {
+        *self.entities.add(archetype_index)
+    }
+}
+
+impl<T: Component> WorldQuery for &T {
+    type Fetch = ReadFetch<T>;
+    type State = ReadState<T>;
+}
+
+pub struct ReadState<T> {
+    component_id: ComponentId,
+    storage_type: StorageType,
+    // NOTE: PhantomData<fn()-> T> gives this safe Send/Sync impls
```

Comment:
An extra field can only remove auto trait implementations. Not add them.

Reply:
`fn()->T` is always `Send` and `Sync`.

Reply:
Then this whole field is useless as adding fields can only remove auto trait impls and not add them.

Reply:
Hmm this is needed in a few places where type parameters might not be send/sync but the type must always be send/sync (`NonSendState<T>`, `NonSendMutState<T>`, and `FunctionSystem<In, Out, Param, Marker, F>`). But I did use this in a number of places where it isn't removed (including this one). I'll remove it in favor of `PhantomData<T>` wherever possible.